Auth logic

DedovaA opened the pull request:
implemented AuthScreen logic

DedovaA committed:
implemented switch button logic

DedovaA committed:
implemented AuthScreen logic

DedovaA committed:
implemented AuthScreen logic

Nikolaiko commented on app/src/main/java/com/example/petproject/AuthViewModel.kt:
```diff
@@ -2,17 +2,147 @@ package com.example.petproject
+     * authScreenType (login / registration)
+     */
+    private val _authFormType = MutableLiveData<AuthFormType>()
+    val authFormType = _authFormType
```
Если ты хочешь ограничить "урезать" возможности свойства с MutableLiveData до LiveData, то нужно явно указать тип публичного свойства :
val authFormType: LiveData<AuthFormType> = _authFormType
иначе котлин автовыведением присвоит authFormType такой же тип данных.

DedovaA replied:
поправила

Nikolaiko commented on app/src/main/java/com/example/petproject/AuthViewModel.kt:
```diff
@@ -2,17 +2,147 @@ package com.example.petproject
+     * "Войти" Button-> onClick
+     */
+    fun loginValidation() {
+        clearScreenStates()
```
Вспомни опыт прошлых приложений, разве после неправильного ввода у тебя отчищались введенные данные? Нет. Это плохой подход. 
Плюс поведение и логика описывается в usecases проекта, там нет ничего про то, что поля должны отчищаться.

DedovaA replied:
убрала из проверок fun loginValidation() и fun registerValidation(). оставила сброс данных только при переключении экрана с "войти" на "регистрацию"

Nikolaiko commented on app/src/main/java/com/example/petproject/AuthViewModel.kt:
```diff
@@ -2,17 +2,147 @@ package com.example.petproject
+        }
+        setNameIsValid(nameState)
+
+        val emailState = when {
```
использование when - хорошо.

Nikolaiko commented on app/src/main/java/com/example/petproject/AuthViewModel.kt:
```diff
@@ -2,17 +2,147 @@ package com.example.petproject
+    }
+
+    private val _email =  MutableLiveData<String>()
+    val email = _email
```
1. Тоже самое с ограничением типа полей.
2. Обычно все-таки объявление полей класса делаются в одном месте сразу. В начале, а потом методы.

DedovaA replied:
поправила

Nikolaiko commented on app/src/main/java/com/example/petproject/AuthViewModel.kt:
```diff
@@ -2,17 +2,147 @@ package com.example.petproject
+    private val _emailIsValid =  MutableLiveData<Boolean>()
+    val emailIsValid = _emailIsValid
+    fun setEmailIsValid(state:Boolean){
+        _emailIsValid.postValue(state)
```
Тут просто подумай еще раз, есть ли необходимость выделять одну строчку в отдельную функцию.
Если решишь, что да - оставляй.

DedovaA replied:
тут не поняла

Nikolaiko replied:
Это я не до конца понял код сначала, все ок.

Nikolaiko commented on app/src/main/java/com/example/petproject/loginScreen/AuthorizationScreen.kt:
```diff
@@ -7,17 +7,26 @@ import androidx.compose.ui.Alignment
 @Composable
 fun AuthorizationScreen(viewModel: AuthViewModel) {
-    val authForm = viewModel.state.observeAsState(AuthFormType.login)
+    val authForm = viewModel.authFormType.observeAsState(AuthFormType.login)
```
Хорошо. Вьюха получилась аккуратной.

Nikolaiko commented on app/src/main/java/com/example/petproject/loginScreen/LoginScreen.kt:
```diff
@@ -1,53 +1,31 @@
+    passwordValid: Boolean,
+    callbackLoginValid: AuthCheckCallback
+) {
+    LoginForm(callbackEmail, callbackPassword, email, password, emailValid, passwordValid)
```
Я рекомендую, если параметров больше трех :
- писать их в столбик
- указывать их имена

DedovaA replied:
в столбик написала, что значит указывать их имена не поняла)

Nikolaiko commented on app/src/main/java/com/example/petproject/utils/UserInputValidation.kt:
```diff
@@ -0,0 +1,19 @@
+}
+
+fun isPasswordMatch(password: String, passConfirm: String) = password == passConfirm
+
```
Тут две функции : isPasswordMatch и isEmptyField - просто однострочники. Тоже, просто еще раз подумай, нужны?

Если  да, оставляй.

DedovaA replied:
согласна, что в принципе они бесполезны, я просто для наглядности их вынесла все отдельный файл

Nikolaiko commented on app/src/main/java/com/example/petproject/viewComponents/forms/LoginForm.kt:
```diff
@@ -23,10 +30,20 @@ fun LoginForm() {
+                    else -> ErrorMessage(stringResource(R.string.error_message_email))
+                }
+                else -> ErrorMessage(stringResource(R.string.empty_string))
+            }
```
1. Если у тебя известное количество вариантов в when, то у последнего лучше явно писать значение варианта : 
true -> 
false ->

а не : 
true ->
else ->

2. Вот это конструкцию как раз можно вынести в отдельную функцию (тем более, что она используется у тебя ниже в другом классе):

`fun example(emailValid, email): Int {
when(emailValid){
                false -> when(email){
                    "" -> R.string.error_message_empty_field)
                    else -> R.string.error_message_email)
                }
                else -> R.string.empty_string)
            }
}
ErrorMessage(stringResource(example(emailValid, email))
`
Или как-то так.

DedovaA replied:
сделал отдельный файл проверками

Nikolaiko commented on app/src/main/res/values/strings.xml:
```diff
@@ -13,6 +13,7 @@
     <string name="paw_icon_description">paw icon</string>
     <string name="logo_description">logo</string>
     <string name="arrow_icon_description">arrow_icon</string>
+    <string name="empty_string"> </string>
```
Пустую строку мы делаем, но в константах обычно.
const val EMPTY_STRING = ""

DedovaA replied:
не знаю, куда вынести эту константу, сделала для нее отдельный файл

DedovaA committed:
added const val EMPTY_STRING = "" to Const.kt,
Extracted fieldValidation methods to Utils.kt,
Fixed AuthViewModel.kt (added fields encapsulation, fixed loginValidation / registerValidation methods),
Reformat code.

DedovaA committed:
added check for null-values in fields
